src/client/activeobjectmgr.cpp

```diff
@@ -69,21 +66,24 @@ bool ActiveObjectMgr::registerObject(std::unique_ptr<ClientActiveObject> obj)
 	}
 	infostream << "Client::ActiveObjectMgr::registerObject(): "
 			<< "added (id=" << obj->getId() << ")" << std::endl;
-	m_active_objects.put(obj->getId(), obj);
+	m_active_objects.put(obj->getId(), std::move(obj));
 	return true;
 }
 
 void ActiveObjectMgr::removeObject(u16 id)
 {
 	verbosestream << "Client::ActiveObjectMgr::removeObject(): "
 			<< "id=" << id << std::endl;
-	auto it = m_active_objects.find(id);
-	if (it == m_active_objects.end()) {
+	auto &it = m_active_objects.get(id);
+	if (!it) {
 		infostream << "Client::ActiveObjectMgr::removeObject(): "
 				<< "id=" << id << " not found" << std::endl;
 		return;
 	}
 
+	// the map doesn't like this but we're about to delete it anyway
+	auto &unsafe_it = const_cast<std::unique_ptr<ClientActiveObject>&>(it);
+	std::unique_ptr<ClientActiveObject> obj = std::move(unsafe_it);
 	m_active_objects.remove(id);
```

Comment:
Can't you add a function `V take(const K &)` to move out a value?

Reply:
This is a good idea however there's still a small difference.
The server code deletes the object *while* it still in the map (so I think `get(id)` would still return something¹) and then removes it from the map.
Using `take` would remove the entry from the map and *then* destruct it.
To me that also sounds much safer but do we rely on the difference anywhere?

¹: not sure if that returns a partially destructed object or a nullptr

Reply:
>The server code deletes the object while it still in the map (so I think get(id) would still return something¹) and then removes it from the map.

You're referring to the code with the `Delete the obj before erasing, as the destructor may indirectly access m_active_objects.` comment, right?

For explanation of this comment: It was meant to refer to reentrance problems if a destructor called by `erase` calls a function on `m_active_objects`. Do *not* read it as "the destructor may indirectly access *objects inside* `m_active_objects`".

Before #13560, the id was erased before the object was deleted. This is similar to moving the object out of the container.
I don't know why I did it the other way. It looks like it just ended up like this due to iterative changes after reviews.

>¹: not sure if that returns a partially destructed object or a nullptr

According to `unique_ptr::reset` doc, the ptr is first set to nullptr, and then the object is deleted.
`getActiveObject` handles nullptrs like non-existent objs, so it's fine either way.

Anyway, the way this PR does it now, with taking out first, is good. :+1:

Reply:
I see, looks like I was overinterpreting it.